distributed/spill.py

```diff
@@ -1,39 +1,81 @@
 from __future__ import annotations
 
+import logging
 from collections.abc import Hashable, Mapping
+from distutils.version import LooseVersion
 from functools import partial
-from typing import Any
+from typing import TYPE_CHECKING, Any
 
-from zict import Buffer, File, Func
+import zict
+
+if TYPE_CHECKING:
+    from typing_extensions import Literal
 
 from .protocol import deserialize_bytes, serialize_bytelist
 from .sizeof import safe_sizeof
 
+logger = logging.getLogger(__name__)
+
 
-class SpillBuffer(Buffer):
+class SpillBuffer(zict.Buffer):
     """MutableMapping that automatically spills out dask key/value pairs to disk when
-    the total size of the stored data exceeds the target
+    the total size of the stored data exceeds the target. If max_spill is provided the
+    key/value pairs won't be spilled once this threshold has been reached.
+
+    parameter
     """
 
-    spilled_by_key: dict[Hashable, int]
-    spilled_total: int
+    def __init__(
+        self, spill_directory: str, target: int, max_spill: int | Literal[False] = False
+    ):
+        if max_spill is not False and LooseVersion(zict.__version__) <= "2.0.0":
+            raise ValueError("zict > 2.0.0 required to set max_weight")
 
-    def __init__(self, spill_directory: str, target: int):
-        self.spilled_by_key = {}
-        self.spilled_total = 0
-        storage = Func(
-            partial(serialize_bytelist, on_error="raise"),
-            deserialize_bytes,
-            File(spill_directory),
-        )
         super().__init__(
-            {},
-            storage,
-            target,
-            weight=self._weight,
-            fast_to_slow_callbacks=[self._on_evict],
-            slow_to_fast_callbacks=[self._on_retrieve],
+            fast={},
+            slow=Slow(spill_directory, max_spill),
+            n=target,
+            weight=_in_memory_weight,
         )
+        self.logged_keys: list[str] = []  # list to track logged keys to avoid spaming
+
+    def __setitem__(self, key, value):
+        try:
+            super().__setitem__(key, value)
+        except MaxSpillExceeded as e:
+            key_e = e.args[0]  # otherwise it returns (key_e, )
+            # key is in self.fast; no keys have been lost on eviction
+            # Note: requires zict > 2.0
+            # TODO don't spam the log file with hundreds of messages per second
+            if key_e not in self.logged_keys:
+                logger.warning(
+                    "Spill file on disk reached capacity; keeping data in memory"
+                )
+                self.logged_keys.append(key_e)
```

Comment:
For managing the logging problem, I would suggest to deal with this via stdlib logging filters, see https://docs.python.org/3/library/logging.html#filter-objects

For instance, this one would only log the same message only X seconds and would attach a summary.

```python
from collections import defaultdict
import logging
from distributed.utils import time
from dask.utils import parse_timedelta

class SuppressLoggingTime(logging.Filter):
    def __init__(self, log_every="2s"):
        self.last_logged = {}
        self.counters = defaultdict(int)
        self.log_every = parse_timedelta(log_every)
        
    def filter(self, record):
        msg = record.msg
        created = record.created
        self.counters[msg] += 1
        last_logged = self.last_logged.get(msg)
        if last_logged is None or created - last_logged > self.log_every:
            self.last_logged[msg] = created
            counter, self.counters[msg] = self.counters[msg], 0
            if counter > 1:
                record.msg += f"\n -- {counter - 1} similar messages since last log was emitted {created - last_logged:.2f}s ago."
            return True
        return False

# usage

logger = logging.getLogger("test")

# First message is logged as is
logger.critical("foo")
logger.critical("foo")
logger.critical("foo")

_filter = SuppressLoggingTime()
logger.addFilter(_filter)
# Now we only receive a log every 2s
while True:
    logger.critical("foo")

```
![image](https://user-images.githubusercontent.com/8629629/150114332-fb6828c9-f0af-4a2e-94c1-f0e008053080.png)

This should be better performing and would remove any logging logic from the application code.


---

at the very least `logged_keys` should be a set and not a list. We wouldn't want logging logic to introduce a performance problem (`is in` can become expensive for large lists) . I'm also a bit concerned about an every growing collection regardless of whether this is a set, list, dict. My solution has a similar problem.